Improve Split accuracy.

xStrom opened the pull request:
This PR includes a slew of changes to the `Split` widget.

* All layout and drawing is now quantized to integers as a continuation of work in #669. This means that single pixel lines never bleed out to side pixels. This applies to both the inner children and the split bar.
* Adds `min_splitter_area(f64)` method which basically acts as padding around the split bar. This padding still reacts to mouse events. As a result it is possible to use extremely narrow split bars while still having a reasonable area for the user to click on. Fixes #641.
* No longer forgets the `split_point` chosen by the user. The widget still modifies the split point when the size constraints require it (e.g. when resizing the window to extremely small) however now the user chosen split point is restored as soon as the constraints allow it.
* The meaning of `splitter_size` has slightly changed, to a less surprising interpretation.
    * Previously a splitter size of 10 with a solid fill meant:
     4px wide solid fill with 3px padding on each side.
    With the fill width formula being: size - 2*((size/3).floor())
    Now a 10 with a solid fill means a 10px wide solid fill.
    The new `min_splitter_area` can be used to add padding.
    * Previously a splitter size of 10 with a line style meant:
    2px pad | ~2px line | 2px pad | ~2px line | 2px pad
    With 1px lines being centered (size/3).floor() pixels away from edges.
    Now a 10 with a line style means: 3px line | 4px pad | 3px line
    Padding at edges can be added optionally with `min_splitter_area`.
* Renamed a few internal functions and added a few lines of documentation for better readability.

Generally speaking the `Split` widget still has issues, which were there before and aren't addressed here, e.g. children still being drawn when they have zero sized layout. I'll probably tackle this at least partially in a future PR.

cmyr commented:
@Zarenor are you around? Any interest in taking a look at this?

cmyr commented on druid/src/widget/split.rs:
```diff
@@ -88,38 +98,89 @@ impl<T> Split<T> {
+    /// This can be useful when you want to use a very narrow visual splitter,
+    /// but don't want to sacrifice user experience.
+    ///
+    /// The default minimum splitter area is `6.0`.
```
I would include a note that this panics if the size is less than 0.0.

cmyr commented on druid/src/widget/split.rs:
```diff
@@ -44,42 +46,50 @@ impl<T> Split<T> {
-    /// Set the width of the splitter bar, in pixels
-    /// The value must be positive or zero
+
+    /// Set the width of the splitter bar.
```
```suggestion
    /// Builder-style method to set the width of the splitter bar.
```

Zarenor replied:
Should we start adding methods that allow mutation of this kind of thing at runtime (so that this distinction is meaningful)?

I think the reason we haven't thus far is the type-erased nature of children contained by built-in container widgets? But it feels like we're moving in a direction where it's acceptable to work around that or downcast out of type-erasure. So the distinction could *become* meaningful.

cmyr replied:
I'm just trying to make this a convention in the docs for any builder-style methods, since we have decided to forgo having actual builders.

For mutating methods, my feeling is "it depends". Some cases feel really obvious (changing label color) but others I'm less clear about. I don't think we need to add them everywhere; we can see what use dictates.

cmyr commented on druid/src/widget/split.rs:
```diff
@@ -88,38 +98,89 @@ impl<T> Split<T> {
         self
     }
+
+    /// Set the minimum width of the splitter area.
```
```suggestion
    /// Builder-style method to set the minimum width of the splitter area.
```

cmyr commented on druid/src/widget/split.rs:
```diff
@@ -88,38 +98,89 @@ impl<T> Split<T> {
     }
-    fn splitter_hit_test(&self, size: Size, mouse_pos: Point) -> bool {
+
+    /// Returns the width of the area of mouse hit detection.
```
'width' is kind of confusing for a vertical splitter, but whatever 🤷‍♂

Zarenor replied:
'breadth', so we're not implying directionality (vs 'width' or 'height')?

Zarenor replied:
I'd used 'size' before, because I hadn't been able to come up with 'breadth'

cmyr replied:
I think breadth is probably more confusing than advantageous. I think size is good?

xStrom replied:
Yeah the usage of `width` is a bit confusing. Previously `Split` used both `size` and `width` to refer to this and I didn't spend much time thinking about it. I think it's best to use only one word for this concept and I'll do that update.

As for what word to use, `size` is decent. I can also think of `thickness` and `cross length`. I kind of like `thickness`.

cmyr replied:
whichever you like; I still prefer size I think, thickness has z-axis implications for me?

xStrom replied:
Well there's no z-axis involved with `Split` so that's an unlikely clash. Also [Flutter seems to use thickness](https://api.flutter.dev/flutter/material/Divider/thickness.html).

That said, I went with `size` for now due to its shortness and better fit with `area`.

cmyr commented on druid/src/widget/split.rs:
```diff
@@ -27,8 +27,10 @@ pub struct Split<T> {
+    split_point_chosen: f64,
+    split_point_effective: f64,
     splitter_size: f64,
+    min_splitter_area: f64,
```
Maybe this should be an `Option<f64>`, so that we can default to using the splitter size when this isn't passed explicitly? Maybe?

I'm not sure. I'm just thinking of in the example where you set the splitter size to 3, should we keep the same default min_splitter_area? Maybe...

xStrom replied:
I think having it set by default with some reasonable value is best. The main reason this exists is to allow extremely narrow splitter bars without sacrificing the user experience by requiring them to pixel hunt. So if someone wants a 3px bar, or even a 1px bar, they can easily do so but druid will default to a 6px bar area. If they want the area to also be 1px then they can set the minimum area to e.g. 0px.

Zarenor commented on druid/src/widget/split.rs:
```diff
@@ -88,38 +98,89 @@ impl<T> Split<T> {
+        self.splitter_size.max(self.min_splitter_area)
+    }
+
+    /// Returns the padding width added to the splitter on both sides.
```
```suggestion
    /// Returns the padding width added to the splitter on each side.
```
I think the 'each side' phrasing makes more clear that we're dividing the difference by two - we *should* add this to each side, not redivide.

cmyr replied:
agreed.

Zarenor commented on druid/src/widget/split.rs:
```diff
@@ -88,38 +98,89 @@ impl<T> Split<T> {
 
-        let min_offset = (self.splitter_size * 0.5).min(5.0);
-        let mut min_limit = self.min_size.max(min_offset);
-        let mut max_limit = (size_in_splitted_direction - self.min_size.max(min_offset)).max(0.0);
+        let mut min_limit = self.min_size;
+        let mut max_limit = (size_in_split_direction - min_limit).max(0.0);
```
I see we've removed the absolute '5.0' minimum. This was originally added to 'fix' the possibility that the splitter would get 'lost' in the edge of it's container - or that nested splitters could collide.
I'm okay with removing it if the behavior is still acceptable in these extrema, but it's something worth checking for.

xStrom replied:
This minimum isn't needed with the new logic, because the splitter bar is guaranteed to always be visible. This specific function calculates the min/max limits given the provided size and doesn't need to know about the splitter bar at all. The bar logic is elsewhere.

Zarenor commented:
I'm sorry I didn't see this earlier. I added a few more comments here.

I'm really glad to see the work on quantizing move forward, and I appreciate the other improvements made.

Looks great!

xStrom commented:
I addressed the feedback and made some other improvements.

* Improved documentation, including unifying language for builder methods.
* Renamed a bunch of variables and methods for consistency and reducing confusion, e.g.:
    * What was known previously as **splitter** is now the **splitter bar** to make it clear it's different from `Split` which is the whole widget. So **splitter bar** is how it's referenced in documentation, but actual methods are just `bar` for shortness.
    * The width/breadth/thickness of the splitter bar is now always called `size`.
    * `fill_splitter_handle` to `solid_bar` to use the new `bar` nomenclature and to make it more concise.
    * `split_direction` is now `split_axis` and this has some functional changes too, which I describe below.
* Swapped the meaning of the `Split::new(axis)`, `Split::horizontal()`, and `Split::vertical()` constructors. Previously it meant that the splitter bar is horizontal/vertical, but now it means that the horizontal/vertical axis is split in half. I think this reads better and this also matches what other frameworks are doing, e.g. [Qt](https://doc.qt.io/qt-5/qsplitter.html#orientation-prop) and [Metro Components](https://metroui.org.ua/splitter.html).
* The effective split point calculation now properly takes the bar area into account.
* The minimum split child size, bar size, and bar area is now rounded up to integers.
* Fixed the check for infinite height.
* Reduced code size a bit by some refactoring.

xStrom commented:
Based on the discussion in [zulip](https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/split.20vertical.2Fhorizontal.20confusion) I renamed the constructor helpers from vertical/horizontal to row/cols.
```rust
Split::columns(child1, child2) // left and right
Split::rows(child1, child2)    // top and bottom
```
